README.md
# arrival_closed

context/history.md
---
type: context/history
name: Tales from the past
world: Guemenos
campaign: Arrival
tags: 
---

# Tales from the past

## The world known as Guemenos

In the darkness of a forgotten universe, the world of Guemenos was born from the hand of Nek, a deity who crafted its very foundations. With its inception, Nek gifted the realm with an intricate tapestry of energy known as Theurgy, an otherworldly power that pulsed through the fabric of existence.

Intriguingly, the powers of Theurgy took on a life of their own, self-organising and imbuing Guemenos with life, birthing an array of creatures that spanned its domains. Among these enigmatic beings were the Weavers, ethereal races forged from the very essence of Theurgy itself. They embodied the raw potential of this arcane energy, possessing a mastery that surpassed mortal comprehension.

Over the course of centuries, the Weavers delved into the mysteries of Theurgy, researching deeper into its enigmatic depths. Through their relentless work, they unlocked the secrets of Theotechnical engines, wondrous devices that harnessed and channeled the boundless power of Theurgy. With each invention, Guemenos flourished, and its inhabitants revelled in the marvels that these devices wrought.

### Of Weavers and Angels

As is often the way with great power, discord brewed amidst the Weavers. Their divergent beliefs clashed, and the realm was consumed by the ravages of imperial wars that scarred the land for ages untold. Seeking resolution, the Weavers convened in a grand council, their intentions set upon parlaying with the enigmatic Nek himself, hoping to find a common ground and end the strife that plagued their world.

In their anticipation, the Weavers ventured to Nek's halls, expecting to encounter the deity himself. To their dismay, they were instead greeted by the ethereal presence of Angels, emissaries of Nek. Betrayal dripped from the air as the Weavers realised that Nek had shunned their entreaties, substituting his own celestial messengers in his stead.

Enraged by this apparent treachery, the Weavers vowed to rise above Nek and his celestial minions. They intensified their efforts, seeking to augment their Theotechnical engines to surpass the power of Nek and the Angels. Alas, their ambitions would be dashed against the shores of fate, for the Angels, envious of the world Nek had woven for the Weavers and lesser races, devised a plan to seize Guemenos for themselves.

Eons of bloodshed ensued as the Angels waged relentless war against the Weavers. The heavens trembled with their clash, and the inevitable came to pass—the Angels emerged victorious. They claimed Guemenos as their own, declaring themselves The Edraihil, the self-proclaimed sons and daughters of Nek.

Banished to the depths of the world they once called home, the Weavers and their peoples retreated, their powers simmering in the shadows. For millennia, their knowledge and strength melded in the crucible of isolation, ultimately culminating in the construction of the Kyri. This powerful forged pantheon stood as a testament to their tenacity, a bastion of power to reclaim what had been taken.

As the ages unfolded, most of the Edraihil gradually waned in power, their once celestial forms degraded into mere shadows of their former glory. Yet, a select few endured, ascending to blessed stature, their essence resonating with the divine. These remnants of the Edraihil persisted as powerful beings, revered and feared by those who dared cross their path. A chapter in the grand tapestry of Guemenos had closed, while another, steeped in dark intrigue and ancient grudges, unfurled in the depths below.

### The rule of the Edraihil

The Edrahil, their power solidified and their dominion unchallenged, reshaped Guemenos in the image of their celestial progenitor, Nek. With intricate incantations and unfathomable rituals, they twisted the very fabric of reality, moulding the land to reflect their desires. Mountains rose like jagged fangs, rivers twisted and turned in unnerving patterns, and forests whispered secrets that only the chosen few could decipher.

In their divine fervour, the Edrahil bestowed life upon new creatures, imbuing them with aspects that paid homage to their own divine essence. From the writhing depths of the oceans to the desolate expanse of the deserts, monstrous beings were a grotesque reflection of their creators, bearing distorted features and unearthly forms. These abominations roamed, forever serving as a testament to the Edrahil's power and their mastery over creation itself.

For millennia, Guemenos thrived under the rule of the three eldest Edrahil, each one presiding over a distinct realm within the twisted tapestry of the world. The land teemed with life, albeit with twisted accents, and the inhabitants lived in a delicate balance between awe and terror. The Edrahil were revered as gods, their presence feared and worshipped by the lesser beings who lived in the three realms.

But deep within the heart of Guemenos, a foreboding darkness stirred. The cosmic fabric that held this twisted realm together quivered with an otherworldly anticipation, whispering of a lurking menace that threatened to shatter the fragile equilibrium. Shadows loomed larger, madness dripped like poison from the air, and whispers of an ancient, eldritch force began to echo through the ethereal halls of Guemenos.

#### Iopothae


rules/_aux/civTech.R
library(rjson)
library(dplyr)
library(stringr)
library(jsonlite)

# import data
techTree <- rjson::fromJSON(file = 'rules/_aux/civTechTree.json')

# clean data
techTree <- lapply(techTree, function(tech) {
  if (str_detect(tech$tech_name, '\\(')) {
    tech$tech_name <- sub(" \\(Civ5)", "", tech$tech_name)
  }
  return (tech)
})

# create simple table for reference
techTable <- lapply(techTree, function(tech) {
  data.frame(tech_id = tech$tech_id,
             tech_name = str_to_lower(tech$tech_name),
             era = tech$era,
             cost = tech$cost,
             units = toString(tech$units_enabled))
})
techTable <- do.call(rbind, techTable)
names(techTree) <- techTable$tech_name

# summarise cost per era, and calculate yearly costs per era
costPerEra <- techTable %>%
  group_by(era) %>%
  summarise(totCost = sum(cost)) %>%
  arrange(totCost) %>%
  mutate(cumCost = cumsum(totCost),
         yearsInEra = c(3000, 1500, 850, 375, 165, 55, 50, 25),
         cumYears = cumsum(yearsInEra),
         progrPerYear = totCost / yearsInEra)

techInfoTbl <- costPerEra %>%
  mutate(era_id = 1:8) %>%
  select(era_id, era, yearsInEra)
  

techInfoTbl <- techInfoTbl %>%
  left_join(techTable %>%
              group_by(era) %>%
              summarise(techs = paste0(tech_name, collapse = ', ')),
            by = 'era')

write.csv(techInfoTbl, 'rules/_aux/techInfoTbl.csv', row.names = FALSE)



npcs/_generators/npcGen.R
# ---
# title: "NPC Generator"
# date: "2023-08-23"
# ---
  
# libraries
library(dplyr)
library(yaml)
library(EnvStats)

# import data
dataLst <- read_yaml('arrival/npcs/_generators/lists01.md')

# define list type based on header information
npc <- lapply(names(dataLst), function (entry) {
  # sample entry on a sub list
  txts <- sample(dataLst[[entry]], 1)
  
  # if sub list contains more details, retrieve those
  if (is.list(txts)) {
    
    # correct for case of entry without detail in sub list where other
    # entries had details
    if (is.null(names(txts[[1]]))) {
      result <- paste0('- **', entry, '**: ', txts[[1]][[1]], '\n')
    } else {
      
      # retrieve sub list entry details
      result1 <- paste0('- **', entry, '**: ', names(txts[[1]]), '\n    - ')
      result2 <- paste0(txts[[1]][[1]], collapse = '\n    - ')
      result <- paste0(result1, result2, '\n') 
    }
    
    # retrieve sub list with simple entries
  } else {
    result <- paste0('- **', entry, '**: ', txts, '\n')
  }
  
  # define returned entry
  return (result)
}) 

# print for md
cat('', paste0(unlist(npc)))


locations/encounters/encounterGen2.R
# ---
# title: "Generic Encounter Generator"
# date: "2023-09-25"
# ---
  
# libraries
library(dplyr)
library(yaml)
library(EnvStats)

# script to evaluate encounter rank
if (FALSE) {
  n <- 10000
  
  f1n <- round(rtri(n, 1, 8, 3.5))
  f2n <- round(rtri(n, 0.5, 1.5, 1))
  f3n <- round(rtri(n, 0, 2, 0.5))
  f4n <- rep(1, n)
  f5n <- round(rtri(n, 0, 2, 1))
  
  hist(f1n + f2n + f3n + f4n + f5n)
  mean(f1n + f2n + f3n + f4n + f5n)
  
  f1r <- 1
  f2r <- 3
  f3r <- 5
  f4r <- 7
  f5r <- 10
  
  hist(f1n*f1r + f2n*f2r + f3n*f3r + f4n*f4r + f5n*f5r)
  mean(f1n*f1r + f2n*f2r + f3n*f3r + f4n*f4r + f5n*f5r)
}

# function to generate the encounter code for obsidian
genEncCode <- function (entry) {
  code <- '```encounter\ncreatures:\n'
  for (i in seq_along(entry)) {
    args <- as.numeric(strsplit(names(entry[[i]]), '-')[[1]])
    
    if (length(args) == 1) {
      numberEnc <- args
    } else {
      numberEnc <- round(rtri(1, min = args[1], max = args[2], mode = args[3]))
    }
    
    code <- paste0(code, ' - ', numberEnc, ': ', as.character(entry[[i]]), '\n')
  }
  
  code <- paste0(code, '```')
  return (code)
}

# function to return the reaction based on roll
genReacRoll <- function (mod = 0) {
  reacRoll <- sample(1:6, 1) + sample(1:6, 1) + mod
  reaction <- case_when(
    reacRoll == 2 ~ '2: As aggressively hostile as the situation allows',
    reacRoll <= 5 ~ '3-5: More unfriendly and hostile than they’d be expected to be in the given situation',
    reacRoll <= 8 ~ '6-8: As predictably hostile or friendly as they’d usually be in this situation',
    reacRoll <= 11 ~ '9-11: More friendly and benign than you’d expect them to be, given the circumstances',
    reacRoll == 12 ~ '12+: As friendly and helpful as their nature and the situation permits them to be',
  )
  return (reaction)
}

# import data
dataLst <- read_yaml('arrival/locations/encounters/list_linhe.md')

# user inputs
time <- 'night'
rank <- 'r1'
reacMod <- 0

# define distance and encounter type
distance <- (sample(1:6, 1) + sample(1:6, 1) + sample(1:6, 1) + 1) * 5
type <- sample(dataLst$type, 1)
reaction <- genReacRoll(reacMod)

# define list type based on header information
if (type == 'effects') {
  lstName <- type
} else {
  lstName <- paste(type, time, rank, sep = '_')
  lstActName <- paste0(type, '_activities')
}

# store header
header <- paste0('\nIt is ', time, ' in a hex of rank ', rank,
                 '\nThere are ', type, ', ', distance, ' meters away')

# encounter code
if (type == 'effects') {
  encCode <- sample(dataLst$effects, 1)
} else {
  lstIndex <- seq_along(dataLst[[lstName]])
  lstChoice <- sample(lstIndex, 1)
  actChoice <- sample(dataLst[[lstActName]], 1)
  
  encCode <- paste0('  ***  ', names(dataLst[[lstName]][[lstChoice]]), '  ***  ',
                    '\nThey are ', actChoice,
                    '\nTheir reaction is ', reaction, '\n\n',
                    genEncCode(dataLst[[lstName]][[lstChoice]][[1]]))
}

# print encounter
cat(paste0(header, '\n\n',
           encCode))


campaign/siliconSamba/districts/encounters/encounterGenSS.R
# ---
# title: "Silicon Samba City Encounters"
# date: "2023-08-03"
# ---
  
# libraries
library(dplyr)
library(yaml)
library(EnvStats)

# function to return the reaction based on roll
genReacRoll <- function (mod = 0) {
  reacRoll <- sample(1:6, 1) + sample(1:6, 1) + mod
  reaction <- case_when(
    reacRoll == 2 ~ '2: As aggressively hostile as the situation allows',
    reacRoll <= 5 ~ '3-5: More unfriendly and hostile than they’d be expected to be in the given situation',
    reacRoll <= 8 ~ '6-8: As predictably hostile or friendly as they’d usually be in this situation',
    reacRoll <= 11 ~ '9-11: More friendly and benign than you’d expect them to be, given the circumstances',
    reacRoll == 12 ~ '12+: As friendly and helpful as their nature and the situation permits them to be',
  )
  return (reaction)
}

# user inputs
rank <- 0
reacMod <- 0

# define distance, reaction
distance <- (sample(1:6, 1) + sample(1:6, 1) + sample(1:6, 1) + 1) * 5
reaction <- genReacRoll(reacMod)

# define HD of encounter
if (rank == 0) {
  hd_distr <- c(1, 7, 4)
} else {
  hd_total <- rank * 10
  hd_distr <- hd_total + c(-9, 9, 0)
}

hd_enc <- round(rtri(1, min = hd_distr[1],max = hd_distr[2],
                     mode = hd_distr[3]))


# store encounter text
enc_text <- paste0('\nIn a location of rank ', rank,
                   ' the encounter happens at ', distance, ' meters away',
                   '.\nTheir total hit dice is ', hd_enc,
                   '.\nTheir reaction is ', reaction)

# print encounter
cat(enc_text)


locations/encounters/encounterGen.R
# ---
# title: "Wandering Encounters"
# date: "2023-08-03"
# ---
  
# libraries
library(dplyr)
library(yaml)
library(EnvStats)

# script to evaluate encounter rank
if (FALSE) {
  n <- 10000
  
  f1n <- round(rtri(n, 1, 8, 3.5))
  f2n <- round(rtri(n, 0.5, 1.5, 1))
  f3n <- round(rtri(n, 0, 2, 0.5))
  f4n <- rep(1, n)
  f5n <- round(rtri(n, 0, 2, 1))
  
  hist(f1n + f2n + f3n + f4n + f5n)
  mean(f1n + f2n + f3n + f4n + f5n)
  
  f1r <- 1
  f2r <- 3
  f3r <- 5
  f4r <- 7
  f5r <- 10
  
  hist(f1n*f1r + f2n*f2r + f3n*f3r + f4n*f4r + f5n*f5r)
  mean(f1n*f1r + f2n*f2r + f3n*f3r + f4n*f4r + f5n*f5r)
}

# function to generate the encounter code for obsidian
genEncCode <- function (entry) {
  code <- '```encounter\ncreatures:\n'
  for (i in seq_along(entry)) {
    args <- as.numeric(strsplit(names(entry[[i]]), '-')[[1]])
    
    if (length(args) == 1) {
      numberEnc <- args
    } else {
      numberEnc <- round(rtri(1, min = args[1], max = args[2], mode = args[3]))
    }
    
    code <- paste0(code, ' - ', numberEnc, ': ', as.character(entry[[i]]), '\n')
  }
  
  code <- paste0(code, '```')
  return (code)
}

# function to return the reaction based on roll
genReacRoll <- function (mod = 0) {
  reacRoll <- sample(1:6, 1) + sample(1:6, 1) + mod
  reaction <- case_when(
    reacRoll == 2 ~ '2: As aggressively hostile as the situation allows',
    reacRoll <= 5 ~ '3-5: More unfriendly and hostile than they’d be expected to be in the given situation',
    reacRoll <= 8 ~ '6-8: As predictably hostile or friendly as they’d usually be in this situation',
    reacRoll <= 11 ~ '9-11: More friendly and benign than you’d expect them to be, given the circumstances',
    reacRoll == 12 ~ '12+: As friendly and helpful as their nature and the situation permits them to be',
  )
  return (reaction)
}

# import data
dataLst <- read_yaml('arrival/locations/encounters/list_linhe.md')

# user inputs
time <- 'night'
rank <- 'r1'
reacMod <- 0

# define distance
  # arctic, desert, farmland, or grassland | 6d6 × 3m
  # forest, swamp, or woodland             | 2d8 × 3m
  # hills or wastelands                    | 2d10 × 3m
  # jungle                                 | 2d6 × 3m
  # mountains                              | 4d10 × 3m
distance <- (sample(1:6, 1) + sample(1:6, 1) + sample(1:6, 1) + 1) * 5

# define encounter type
type <- sample(dataLst$type, 1)
reaction <- genReacRoll(reacMod)

# define list type based on header information
if (type == 'effects') {
  lstName <- type
} else {
  lstName <- paste(type, time, rank, sep = '_')
  lstActName <- paste0(type, '_activities')
}

# store header
header <- paste0('\nIt is ', time, ' in a hex of rank ', rank,
                 '\nThere are ', type, ', ', distance, ' meters away')

# encounter code
if (type == 'effects') {
  encCode <- sample(dataLst$effects, 1)
} else {
  lstIndex <- seq_along(dataLst[[lstName]])
  lstChoice <- sample(lstIndex, 1)
  actChoice <- sample(dataLst[[lstActName]], 1)
  
  encCode <- paste0('  ***  ', names(dataLst[[lstName]][[lstChoice]]), '  ***  ',
                    '\nThey are ', actChoice,
                    '\nTheir reaction is ', reaction, '\n\n',
                    genEncCode(dataLst[[lstName]][[lstChoice]][[1]]))
}

# print encounter
cat(paste0(header, '\n\n',
           encCode))
